astronomer/providers/sftp/hooks/sftp.py

```diff
@@ -196,6 +129,29 @@ async def list_directory(self, path: str = "") -> list[str] | None:
         except asyncssh.SFTPNoSuchFile:
             return None
 
+    async def read_directory(self, path: str = "") -> Sequence[asyncssh.sftp.SFTPName] | None:
+        """Returns a list of files along with their attributes on the SFTP server at the provided path"""
+        ssh_conn = await self._get_conn()
+        sftp_client = await ssh_conn.start_sftp_client()
+        try:
+            files = await sftp_client.readdir(path)
+            return files
+        except asyncssh.SFTPNoSuchFile:
+            return None
+
+    async def get_files_and_attrs_by_pattern(
+        self, path: str = "", fnmatch_pattern: str = ""
+    ) -> Sequence[asyncssh.sftp.SFTPName]:
+        """
+        Returns the files along with their attributes matching the file pattern (e.g. ``*.pdf``) at the provided path,
+        if one exists. Otherwise, raises an AirflowException to be handled upstream for deferring
+        """
+        files_list = await self.read_directory(path)
+        if files_list is None:
+            raise AirflowException(f"No files at path {path} found...")
```

Comment:
Nit: repr to quote it

```suggestion
            raise AirflowException(f"No files at path {path!r} found...")
```

Reply:
Wonder if this should be 
```suggestion
            raise NotADirectoryError(f"No files at path {path!r} found...")
```

There is no reason it has to be AirflowException -- any exception is valid